Gke deployment

ziqizh opened the pull request:
## Changes
- Containerized servers by creating Docker files.
- Added gRPC calls to register supplier address and vendor addresses.
- Allowed the client to specify Finder address.
- 

## How To Run
1.  `make supplyfinder-client`
2. For the first time running the client, use `./supplyfinder-client -s [Supplier server address] -f [Finder server address]`
e.g. `./supplyfinder-client -f 35.194.10.42:50051 -s 35.232.222.57:50052`
3. The client will ask you to register vendor information. If you has registered before, no need to register it again. 
e.g. 
```
34.72.6.187:50053
Wegmans
NY
35.232.183.141:50053
Kroger
MI
35.188.143.126:50053
Trader Joe's
PA
```
4. The client will prompt you to input your requests.

Commit:
containerize servers

Commit:
containerize servers

Commit:
containerize servers

Commit:
change streaming function

Commit:
change streaming function

Commit:
containerized servers

Commit:
add deploy script

Commit:
remove destructor and reader wrapper

Commit:
fix bugs

Commit:
update yaml

Commit:
add address updating functionality

Commit:
add services

Commit:
add comments

nholbrook commented on supplier/supplier.cc:
```diff
@@ -42,21 +44,12 @@ class SupplierServiceImpl final : public Supplier::Service {
   // DB maintains a mapping ID -> vector<VendorInfo*>.
   // Stores pointers to VendorInfos, which are stored in an array
   std::unordered_map<uint32_t, vector<VendorInfo*>> vendor_db_;
```
Why not just use a vector of VendorInfos?

ziqizh replied:
Because when a vector expands it might move the elements to a new address, which will cause segfault since I stored all the old pointers in vendor_db_

nholbrook commented on supplier/supplier.cc:
```diff
@@ -72,6 +65,23 @@ class SupplierServiceImpl final : public Supplier::Service {
+
+  Status AddVendor(ServerContext* context, const VendorInfo* request,
+                   Empty* info) {
+    if (vendor_count_ > 9) {
```
Using a vector for your vendor storage would remove the need for this hard limit of 9 vendors.

ziqizh replied:
Yeah I think there might be better workaround, but for the sake of time, I decide to use a fix number of vendors.

nholbrook commented on vendor/vendor.cc:
```diff
@@ -92,12 +96,13 @@ void RunServer(std::string addr) {
+  int server_number = 1;
   int base_port = 50053;
+
   std::vector<std::thread> threads;
```
This threading is no longer being used, consider removing.

ziqizh replied:
Nice catch! Thanks Nick!

nholbrook reviewed (approved):
Overall looks good. Just a couple small suggestions.

Commit:
fix input bug

jsandler18 commented on api_config.yaml:
```diff
@@ -0,0 +1,21 @@
+title: Supply Finder API
+apis:
+- name: supplyfinder.Finder
+- name: supplyfinder.Vendor
```
I am not too familiar with how GCP Endpoints works, but I am pretty sure you only want to expose the Finder.  The Vendor and Supplier are implementation details.

jsandler18 commented on proto/supplyfinder.proto:
```diff
@@ -13,6 +14,9 @@ service Finder {
   rpc CheckFood (FinderRequest) returns (stream ShopInfo) {}
+
+  // Register the supplier using Supplier Info.
+  rpc UpdateSupplier (SupplierInfo) returns (google.protobuf.Empty) {}
```
RegisterSupplier?

jsandler18 commented on client/client.cc:
```diff
@@ -0,0 +1,195 @@
+    std::cout
+        << "please input server url, name and location, separated by newlines.\n"
+        << "Or enter q to stop." << std::endl;
+    while (getline(std::cin, url) && url != "q") {
```
I am pretty sure GKE has a way to provide addresses directly to where they are needed, e.i. finder would be provided the address and port of the supplier, supplier would be provided with the addresses of the vendors.  I don't know if you have enough time to look into this, but it's something to keep in mind.

jsandler18 commented on finder/finder.cc:
```diff
@@ -123,6 +124,15 @@ long FinderServiceImpl::GetFoodID(const string& food_name) {
 }
 
+Status FinderServiceImpl::UpdateSupplier(grpc::ServerContext* context,
+                                         const supplyfinder::SupplierInfo* info,
```
The fact that the finder talks to the supplier is an implementation detail, and GKE should be able to provide this address directly to your binary on startup.

jsandler18 commented on supplier/supplier.cc:
```diff
@@ -42,21 +45,14 @@ class SupplierServiceImpl final : public Supplier::Service {
   std::unordered_map<uint32_t, vector<VendorInfo*>> vendor_db_;
-  VendorInfo vendor[4];
+  // use C style array to ensure the validity of the address
+  VendorInfo vendor_[10]; 
```
I know you are time constrained, so if you don't have enough time you can ignore this. 

All of this complexity can be avoided if vendor_db_ is an std::vector<std::unique_ptr<VendorInfo>>.  Sure some VendorInfo is duplicated, but they are not very large protos so some duplication is ok.

oulin-coder reviewed (approved):
Took a look mostly at the coding changes, and everything looks good! I think you're ready to start adding Open Census instrumentation.

Commit:
test

Commit:
dynamic address

ziqizh commented:
- Added functionality to dynamically register addresses during deployment. 
- Changed one of the `streaming` to `repeated`.
- Some of the implementation details remain unchanged.
